File: analytics-engine/services/agent-service.ts
/**
 * Agent-based Query Generation Service
 * 
 * This service integrates LangGraph/LangChain SQL agents for improved
 * SQL query generation, especially for large databases.
 * 
 * Based on: https://github.com/Farzad-R/Advanced-QA-and-RAG-Series/tree/main/AgentGraph-Intelligent-Q%26A-and-RAG-System
 */

import { ChatOpenAI } from '@langchain/openai';
import { DataSourceMetadata, AdhocQueryResponse } from '../types';
import { generateAdhocQuery } from './llm-service'; // Fallback to original
import { exploreRelevantSchema as exploreSchemaTool } from '../agents/tools/schema-explorer';
import { validateQuery as validateQueryTool } from '../agents/tools/query-validator';

// Check if agent dependencies are available
let agentAvailable = false;
let ChatOpenAIClass: any;
let createSqlAgent: any;
let SqlToolkit: any;
let SQLDatabase: any;

try {
  // Try to import LangChain dependencies
  // Note: SQLDatabase is Python-only, so TypeScript SQL agent uses Python backend
  ChatOpenAIClass = require('@langchain/openai').ChatOpenAI;
  createSqlAgent = require('langchain/agents').createSqlAgent;
  SqlToolkit = require('langchain/agents/toolkits/sql').SqlToolkit;
  
  // SQLDatabase is not available in TypeScript LangChain (Python-only)
  // We'll use Python backend for SQL agent instead
  SQLDatabase = null;
  
  agentAvailable = true;
  console.log('[AGENT] LangChain dependencies loaded. TypeScript SQL agent will use Python backend.');
} catch (error) {
  console.warn('[AGENT] LangChain dependencies not found. Install with: npm install @langchain/openai langchain @langchain/community');
  console.warn('[AGENT] Falling back to Python backend agent or direct LLM');
  agentAvailable = false;
}

/**
 * Creates a database connection wrapper for SQL agents
 * Uses Python backend for actual database connections via API
 */
async function createDatabaseConnection(connectionString: string): Promise<any> {
  // SQLDatabase is Python-only, so we always use Python backend
  // This function is kept for compatibility but will redirect to Python backend
  throw new Error('TypeScript SQL agent uses Python backend. Use generateQueryWithPythonAgent() instead.');
}

/**
 * Agent-based query generation using LangChain SQL Agent
 * 
 * Note: SQLDatabase is Python-only, so this redirects to Python backend agent.
 * For TypeScript-only SQL agents, use LangGraph agent instead.
 */
export async function generateQueryWithSQLAgent(
  userQuestion: string,
  connectionString: string,
  metadata?: DataSourceMetadata
): Promise<string> {
  // SQLDatabase is Python-only, so redirect to Python backend agent
  console.log('[AGENT] TypeScript SQL agent redirecting to Python backend (SQLDatabase is Python-only)');
  
  const { generateQueryWithPythonAgent } = await import('./python-agent-bridge');
  const result = await generateQueryWithPythonAgent(userQuestion, connectionString, metadata);
  return result.query_content;
}

/**
 * Enhanced query generation with agent fallback
 * 
 * Uses agent-based approach if available and appropriate,
 * falls back to direct LLM if agent fails or not available
 */
export async function generateAdhocQueryEnhanced(
  userQuestion: string,
  metadata: DataSourceMetadata,
  connectionString?: string,
  useAgent: boolean = false
): Promise<AdhocQueryResponse> {
  // Check if we should use agent-based approach
  const shouldUseAgent = useAgent && 
                         agentAvailable && 
                         metadata.source_type === 'SQL_DB' && 
                         connectionString;

  if (shouldUseAgent) {
    try {
      console.log('[AGENT] Using agent-based query generation');
      
      // Generate query using SQL agent
      const query = await generateQueryWithSQLAgent(
        userQuestion,
        connectionString!,
        metadata
      );

      // Generate insight summary using LLM
      const insightPrompt = `Explain what this SQL query does and what insights it provides:\n\n${query}\n\nQuestion: ${userQuestion}`;
      
      const { default: OpenAI } = await import('openai');
      const openai = new OpenAI({
        apiKey: process.env.OPENAI_API_KEY,
      });

      const insightResponse = await openai.chat.completions.create({
        model: process.env.OPENAI_MODEL || 'gpt-4-turbo-preview',
        messages: [
          {
            role: 'user',
            content: insightPrompt,
          },
        ],
        temperature: 0.3,
        max_tokens: 150,
      });

      return {
        query_type: 'SQL_QUERY',
        query_content: query,
        visualization_type: 'auto',
        insight_summary: insightResponse.choices[0]?.message?.content || 'Query generated successfully',
      };
    } catch (error) {
      console.warn('[AGENT] Agent-based generation failed, falling back to direct LLM:', error);
      // Fall back to original method
      return generateAdhocQuery(userQuestion, metadata);
    }
  }

  // Use original direct LLM approach
  return generateAdhocQuery(userQuestion, metadata);
}

/**
 * Schema exploration tool for large databases
 * 
 * Instead of loading all tables upfront, this explores
 * relevant tables based on the user question
 * 
 * Now uses the schema-explorer tool
 */
export async function exploreRelevantSchema(
  userQuestion: string,
  connectionString: string,
  allTables: string[]
): Promise<DataSourceMetadata> {
  return exploreSchemaTool(userQuestion, connectionString, allTables);
}

/**
 * Query validation and refinement tool
 * 
 * Validates generated queries and can refine them if needed
 * Uses the query-validator tool for validation
 */
export async function validateAndRefineQuery(
  query: string,
  userQuestion: string,
  metadata: DataSourceMetadata,
  maxRefinements: number = 3
): Promise<string> {
  const { default: OpenAI } = await import('openai');
  const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
  });

  let currentQuery = query;
  let refinementCount = 0;

  while (refinementCount < maxRefinements) {
    // Use validator tool for validation
    const validation = await validateQueryTool(currentQuery, userQuestion, metadata);

    if (validation.valid) {
      return currentQuery;
    }

    // Refine query based on errors
    if (refinementCount < maxRefinements - 1) {
      const refinePrompt = `Refine this SQL query to fix the errors.

Original Question: ${userQuestion}
Current Query: ${currentQuery}
Errors: ${validation.errors?.join(', ') || 'Unknown errors'}
Suggestions: ${validation.suggestions || 'None'}

Schema:
${JSON.stringify(metadata, null, 2)}

Generate the corrected SQL query:`;

      const refineResponse = await openai.chat.completions.create({
        model: process.env.OPENAI_MODEL || 'gpt-4-turbo-preview',
        messages: [
          {
            role: 'user',
            content: refinePrompt,
          },
        ],
        temperature: 0.2,
      });

      currentQuery = refineResponse.choices[0]?.message?.content || currentQuery;
      // Clean up query (remove markdown if present)
      currentQuery = currentQuery.replace(/^```sql\s*/i, '').replace(/```\s*$/, '').trim();
      refinementCount++;
    } else {
      // Max refinements reached, return current query
      console.warn('[AGENT] Max refinements reached, returning current query');
      return currentQuery;
    }
  }

  return currentQuery;
}

File: analytics-engine/python-backend/agent_service.py
"""
Agent-based Query Generation Service (Python)
 
This service integrates LangChain SQL agents for improved SQL query generation.
Based on: https://github.com/Farzad-R/Advanced-QA-and-RAG-Series/tree/main/AgentGraph-Intelligent-Q%26A-and-RAG-System
"""

import os
import json
from typing import Dict, List, Optional
from sqlalchemy import create_engine, inspect, text
from langchain.agents import create_sql_agent
from langchain.agents.agent_toolkits import create_sql_agent_executor, SQLDatabaseToolkit
from langchain_openai import ChatOpenAI
from langchain_community.utilities import SQLDatabase
from langchain_core.prompts import ChatPromptTemplate


class SQLAgentService:
    """SQL Agent Service using LangChain"""
    
    def __init__(self):
        """Initialize the SQL Agent Service"""
        self.llm = ChatOpenAI(
            model=os.getenv("OPENAI_MODEL", "gpt-4-turbo-preview"),
            temperature=0,
            openai_api_key=os.getenv("OPENAI_API_KEY"),
        )
    
    def create_database_connection(self, connection_string: str) -> SQLDatabase:
        """
        Create a SQLDatabase connection for LangChain
        
        Args:
            connection_string: Database connection string
            
        Returns:
            SQLDatabase instance
        """
        # Normalize connection string
        normalized_connection_string = self._normalize_connection_string(connection_string)
        
        # Create SQLDatabase instance
        db = SQLDatabase.from_uri(normalized_connection_string)
        return db
    
    def _normalize_connection_string(self, connection_string: str) -> str:
        """
        Normalize connection string to handle special characters
        
        Args:
            connection_string: Raw connection string
            
        Returns:
            Normalized connection string
        """
        # Handle URL encoding if needed
        # This is a simplified version - adjust based on your needs
        return connection_string
    
    def generate_query(
        self,
        user_question: str,
        connection_string: str,
        metadata: Optional[Dict] = None
    ) -> Dict:
        """
        Generate SQL query using LangChain SQL Agent
        
        Args:
            user_question: User's natural language question
            connection_string: Database connection string
            metadata: Optional metadata about the database schema
            
        Returns:
            Dictionary with query and insight_summary
        """
        try:
            # Create database connection
            db = self.create_database_connection(connection_string)
            
            # Create SQL toolkit
            toolkit = SQLDatabaseToolkit(db=db, llm=self.llm)
            
            # Create SQL agent executor
            agent_executor = create_sql_agent_executor(
                llm=self.llm,
                toolkit=toolkit,
                verbose=os.getenv("DEBUG", "false").lower() == "true",
            )
            
            # Execute agent
            result = agent_executor.invoke({
                "input": user_question
            })
            
            # Extract query from result
            query = result.get("output", "")
            
            # Generate insight summary
            insight_prompt = ChatPromptTemplate.from_messages([
                ("system", "You are a data analyst. Explain what this SQL query does and what insights it provides."),
                ("user", f"Query: {query}\n\nQuestion: {user_question}\n\nProvide a brief explanation (max 50 words).")
            ])
            
            insight_chain = insight_prompt | self.llm
            insight_response = insight_chain.invoke({})
            insight_summary = insight_response.content if hasattr(insight_response, 'content') else str(insight_response)
            
            return {
                "query": query,
                "insight_summary": insight_summary,
                "success": True
            }
            
        except Exception as e:
            return {
                "error": str(e),
                "success": False
            }
    
    def explore_schema(
        self,
        user_question: str,
        connection_string: str
    ) -> Dict:
        """
        Explore database schema to find relevant tables for a question
        
        Args:
            user_question: User's question
            connection_string: Database connection string
            
        Returns:
            Dictionary with relevant schema metadata
        """
        try:
            # Create database connection
            db = self.create_database_connection(connection_string)
            
            # Get all table names
            all_tables = db.get_usable_table_names()
            
            # Use LLM to identify relevant tables
            table_selection_prompt = f"""Given this question: "{user_question}"

Available tables: {', '.join(all_tables)}

Which tables are likely needed to answer this question? Return a JSON object with a "tables" array.
Example: {{"tables": ["students", "grades"]}}"""
            
            table_response = self.llm.invoke(table_selection_prompt)
            table_content = table_response.content if hasattr(table_response, 'content') else str(table_response)
            
            # Parse table names
            try:
                # Try to extract JSON from response
                import re
                json_match = re.search(r'\{[^}]+\}', table_content)
                if json_match:
                    table_data = json.loads(json_match.group())
                    relevant_tables = table_data.get("tables", all_tables[:5])
                else:
                    relevant_tables = all_tables[:5]  # Default to first 5
            except:
                relevant_tables = all_tables[:5]
            
            # Get schema for relevant tables only
            tables_metadata = []
            for table_name in relevant_tables:
                if table_name in all_tables:
                    table_info = db.get_table_info_no_throw([table_name])
                    tables_metadata.append({
                        "name": table_name,
                        "info": table_info
                    })
            
            return {
                "source_type": "SQL_DB",
                "tables": tables_metadata,
                "success": True
            }
            
        except Exception as e:
            return {
                "error": str(e),
                "success": False
            }


# Global instance
_agent_service = None

def get_agent_service() -> SQLAgentService:
    """Get or create agent service instance"""
    global _agent_service
    if _agent_service is None:
        _agent_service = SQLAgentService()
    return _agent_service

